src/transport/websocket.ts

```diff
@@ -0,0 +1,444 @@
+          subscription.listeners.onError(new NetworkError('Connection was lost.'));
+        });
+      
+      this.subscriptions.removeAll();
+    });
+    this.socket.addEventListener('close', (event) => {
+      if (!event.wasClean) {
+        this.tryReconnectIfNeeded();
+        return;
+      }
+
+      this.subscriptions
+        .getAllAsArray()
+        .forEach(subscription => {
+          subscription.listeners.onEnd(null);
+        });
+
+      this.subscriptions.removeAll();
+    });
+  }
+
+  private tryReconnectIfNeeded () {
+    if (this.socket.readyState !== WSReadyState.Closed) {
+      return;
+    }
+
+    this.connect();
+  }
+
+  public subscribe(
+    path: string,
+    listeners: SubscriptionListeners,
+    headers: ElementsHeaders,
+    pendingSubID?: number
+  ): Subscription {
+    // If connection was closed, try to reconnect
+    this.tryReconnectIfNeeded();
+
+    // Add subscription to pending if socket is not open
+    if (this.socket.readyState !== WSReadyState.Open) {
+      return new WsSubscription(this, this.pendingSubscriptions.addNew(path, listeners, headers));
+    }
+
+    // Add or select subscription
+    const subID = (!pendingSubID) ?
+      this.subscriptions.addNew(path, listeners, headers) :
+      this.subscriptions.add(pendingSubID, path, listeners, headers);
+
+    this.sendMessage(
+      this.getMessage(
+        SubscribeMessageType,
+        subID,
+        path,
+        headers
+      )
+    );
+
+    return new WsSubscription(this, subID);
+  }
+
+  public unsubscribe(subID: number): void {
+    this.sendMessage(
+      this.getMessage(
+        UnsubscribeMessageType,
+        subID
+      )
+    );
+
+    this.subscriptions.get(subID).listeners.onEnd(null);
+    this.subscriptions.remove(subID);
+  }
+
+  private removeAllSubscribtions () {
+    this.subscriptions.removeAll();
+  }
+
+  private getMessage (
+      messageType: WsMessageType,
+      id: number,
+      path?: string,
+      headers?: ElementsHeaders
+  ): Message {
+    return [
+      messageType,
+      id,
+      path,
+      headers
+    ];
+  }
+
+  private sendMessage (message: Message) {
+    if (this.socket.readyState !== WSReadyState.Open) {
+      return console.warn(`Can't send in "${WSReadyState[this.socket.readyState]}" state`);
+    }
+
+    this.socket.send(JSON.stringify(message));
+  }
+
+  private subscription (subID: number) {
+    return this.subscriptions.get(subID);
+  }
+
+  private recieveMessage (event: any) {
+    this.lastMessageReceivedTimestamp = new Date().getTime();
+
+    // First try to parse event to JSON message.
+    let message: Message;
+    try {
+      message = JSON.parse(event.data);
+    } catch (err) {
+      console.warn(new Error('Message is not valid JSON format'));
+      return;
+    }
+
+    // Validate structure of message
+    const nonValidMessageError = this.validateMessage(message);
+    if (nonValidMessageError) {
+      return console.warn(nonValidMessageError);
```

Comment:
This should result in an exception (or termination of the WebSocket). We're ignoring potentially significant errors otherwise.